add nip-42: authentication (#141)
nostr-protocol/nips: 42.md

This NIP defines a way for clients to authenticate to relays by signing an ephemeral event.

## Motivation

A relay may want to require clients to authenticate to access restricted resources. For example,

  - A relay may request payment or other forms of whitelisting to publish events -- this can naïvely be achieved by limiting publication
    to events signed by the whitelisted key, but with this NIP they may choose to accept any events as long as they are published from an
    authenticated user;
  - A relay may limit access to `kind: 4` DMs to only the parties involved in the chat exchange, and for that it may require authentication
    before clients can query for that kind.
  - A relay may limit subscriptions of any kind to paying users or users whitelisted through any other means, and require authentication.

## Protocol flow

This NIP defines a new message, `AUTH`, which clients can send to relays when they want to authenticate. The message is of the following form:

```
["AUTH", <signed-event>]
```

The signed event is an ephemeral event not meant to be published or queried, it must be of `kind: 22242` and content must be set to the

> WebSocket URL of the relay. `created_at` should be the current time. Example:

Author: @Giszmo and I were discussion about the rationale behind this on the other PR, but you said nothing. Is there a strong reason to have these things specified here?

```json
{
  "id": "...",
  "pubkey": "...",
  "created_at": 1669695536,
  "kind": 22242,
  "tags": [],

>   "content": "wss://relay.example.com/",

Reviewer: just a question: should the trailing slash be enforced to guarantee that the url matches the one the relay expects? if one expects origin (without trailing slash) would that collide with the verification?

Author: I may be missing something, but don't think that is necessary at all, see: https://github.com/nostr-protocol/nips/blob/4a5202646a86ce63a85a50b560d1a2fbce277df6/42.md?plain=1#L67-L68

The relay must decode the URL _anyway_ and check if it is acceptable or not according to hostname, path, querystring etc. It doesn't have to match an exact pattern.

  "sig": "..."
}
```

The client may send an auth message right before performing an action for which it knows authentication will be required -- for example, right
before requesting `kind: 4` chat messages --, or it may do right on connection start or at some other moment it deems best.

Upon receiving a message from an unauthenticated user it can't fulfill without authentication, a relay may choose to notify the client. For

> that it can use a `NOTICE` message with a standard prefix `"restricted: "` that is readable both by humans and machines, for example:

Author: Yes.

> that it can use a `NOTICE` message with a standard prefix `"restricted: "` that is readable both by humans and machines, for example:

Reviewer 2: I don't understand Cameri's comment. Is this resolved? It's marked outdated but I see no REQ or CLOSE in the new diff.

```
["NOTICE", "restricted: we can't serve DMs to unauthenticated users, does your client implement NIP-42?"]
```

or

```
["NOTICE", "restricted: we do not accept events from unauthenticated users, please sign up at https://example.com/"]

> ```

Author: Please tone down.

## Signed Event Verification

To verify `AUTH` messages, relays must ensure:

  - that the `kind` is `22242`

>   - that the event was recently signed (~10 minutes, by `created_at`)

Reviewer 3: can we be more explicit about how long we want authentication to last for to avoid confusion? or are we saying about 10 minutes bc it should be up to the relays how long auth lasts for?

Author: No, this is not for how long it will last. This is for when the event must be created. I think it is somewhat implied that the auth lasts for the duration of the websocket connection, but we can make that more explicit.

Reviewer 3: ah ok that makes a lot more sense. yeah might be good to make it a bit more explicit as that implication went right over my head.

Reviewer 4: Can different websocket connections reuse the same auth message?

Reviewer 3: yes i would think so. as long as it is recent enough. but pretty trivial to sign a new auth message.

>   - that the event was recently signed (~10 minutes, by `created_at`)

Author: Yes, that is written somewhere else above after @monlovesmango's questioned the same thing.

NIP-42
======

Authentication of clients to relays
-----------------------------------

`draft` `optional` `author:Semisol` `author:fiatjaf`

This NIP defines a way for clients to authenticate to relays by signing an ephemeral event.

## Motivation

A relay may want to require clients to authenticate to access restricted resources. For example,

  - A relay may request payment or other forms of whitelisting to publish events -- this can naïvely be achieved by limiting publication
    to events signed by the whitelisted key, but with this NIP they may choose to accept any events as long as they are published from an
    authenticated user;
  - A relay may limit access to `kind: 4` DMs to only the parties involved in the chat exchange, and for that it may require authentication
    before clients can query for that kind.
  - A relay may limit subscriptions of any kind to paying users or users whitelisted through any other means, and require authentication.

## Protocol flow

This NIP defines a new message, `AUTH`, which clients can send to relays when they want to authenticate. The message is of the following form:

```

> ["AUTH", <signed-event>]

Reviewer 2: How do unexpecting relays react to this unknown command? They will probably reply with a notice?

So do I understand correctly that we are using AUTH and not EVENT to avoid this being sent to others who then could use it to auth? Which could only happen if the relay did only occasionally support this nip by maybe having different relay implementations behind some round-robin?

I think we could spare us this new command but tell me please what could go wrong ;)

Author: Yes, you should be able to send this command to unexpecting relays and no harm be done. I don't think they will answer with a notice, most likely they will ignore it, but if they do so be it, it is better than having your auth attempt be broadcasted to the world.

I don't see any advantage of using `EVENT` instead of `AUTH`. The relay will already have to have a very specific custom logic for handling auth events anyway, so I think it makes things clearer to have this handled in a separate place already from the get go.

Author: I think they shouldn't. Mine doesn't. Or maybe mine does, I don't remember, but I think it doesn't.

This NIP defines a way for clients to authenticate to relays by signing an ephemeral event.

## Motivation

A relay may want to require clients to authenticate to access restricted resources. For example,

  - A relay may request payment or other forms of whitelisting to publish events -- this can naïvely be achieved by limiting publication
    to events signed by the whitelisted key, but with this NIP they may choose to accept any events as long as they are published from an
    authenticated user;
  - A relay may limit access to `kind: 4` DMs to only the parties involved in the chat exchange, and for that it may require authentication
    before clients can query for that kind.
  - A relay may limit subscriptions of any kind to paying users or users whitelisted through any other means, and require authentication.

## Definitions

This NIP defines a new message, `AUTH`, which relays can send when they support authentication and clients can send to relays when they want
to authenticate. When sent by relays, the message is of the following form:

```
["AUTH", <challenge-string>]
```

And, when sent by clients, of the following form:

```
["AUTH", <signed-event-json>]
```

> The signed event is an ephemeral event not meant to be published or queried, it must be of `kind: 22242` and it should have at least two tags,
> one for the relay URL and one for the challenge string as received from the relay. `created_at` should be the current time. Example:

Reviewer 5: ```suggestion
The signed event is an ephemeral event not meant to be published or queried, it must be of `kind: 22242` and it should have at least two tags,
one for the relay URL and one for the challenge string as received from the relay.
Relays MUST exclude `kind: 22242` events from being broadcasted to any client.
`created_at` should be the current time. Example:
```

> one for the relay URL and one for the challenge string as received from the relay. `created_at` should be the current time. Example:

Reviewer 6: ```suggestion
one for the relay URL and one for the challenge string as received from the relay. `created_at` should be the current time. Relays should use a maximum acceptable clock skew between its own clock and the client's clock, and to reject signed events with `created_at` timestamps that are outside of this acceptable range. Example:
```

    authenticated user;
  - A relay may limit access to `kind: 4` DMs to only the parties involved in the chat exchange, and for that it may require authentication
    before clients can query for that kind.
  - A relay may limit subscriptions of any kind to paying users or users whitelisted through any other means, and require authentication.

## Definitions

This NIP defines a new message, `AUTH`, which relays can send when they support authentication and clients can send to relays when they want
to authenticate. When sent by relays, the message is of the following form:

```
["AUTH", <challenge-string>]
```

And, when sent by clients, of the following form:

```
["AUTH", <signed-event-json>]
```

The signed event is an ephemeral event not meant to be published or queried, it must be of `kind: 22242` and it should have at least two tags,
one for the relay URL and one for the challenge string as received from the relay.
Relays MUST exclude `kind: 22242` events from being broadcasted to any client.
`created_at` should be the current time. Example:

```json
{
  "id": "...",
  "pubkey": "...",
  "created_at": 1669695536,
  "kind": 22242,
  "tags": [

>     ["relay", "wss://relay.example.com/"],

Reviewer 2: If it's challenge-response, do we really need the relay here?

Author: Yes, otherwise one malicious relay can reuse your auth event with it to login on a different relay.

Reviewer 2: What is this a reply to? My suggestion to drop the commitment to a relay name?

If your relay doesn't re-use challenge nonces, it will associate each nonce with only one websocket connection ever. That websocket connection implicitly commits to the relay already, so naming the relay is not necessary.

> The signed event is an ephemeral event not meant to be published or queried, it must be of `kind: 22242` and it should have at least two tags,
> one for the relay URL and one for the challenge string as received from the relay.
> Relays MUST exclude `kind: 22242` events from being broadcasted to any client.
> `created_at` should be the current time. Example:
> 
> ```json
> {
>   "id": "...",
>   "pubkey": "...",
>   "created_at": 1669695536,
>   "kind": 22242,
>   "tags": [
>     ["relay", "wss://relay.example.com/"],
>     ["challenge", "challengestringhere"]
>   ],
>   "content": "",
>   "sig": "..."
> }

Reviewer 2: If it's challenge-response, do we really need to be so prescriptive about all these details?

```suggestion
The signed event is an ephemeral event not meant to be published or queried. It must have a tag with the challenge string as received from the relay. Example:

```json
{
  "id": "...",
  "pubkey": "...",
  "created_at": 1669695536,
  "kind": 22242,
  "tags": [
    ["challenge", "challengestringhere"]
  ],
  "content": "",
  "sig": "..."
}
```

> The signed event is an ephemeral event not meant to be published or queried, it must be of `kind: 22242` and it should have at least two tags,
> one for the relay URL and one for the challenge string as received from the relay.
> Relays MUST exclude `kind: 22242` events from being broadcasted to any client.
> `created_at` should be the current time. Example:
> 
> ```json
> {
>   "id": "...",
>   "pubkey": "...",
>   "created_at": 1669695536,
>   "kind": 22242,
>   "tags": [
>     ["relay", "wss://relay.example.com/"],
>     ["challenge", "challengestringhere"]
>   ],
>   "content": "",
>   "sig": "..."
> }
> ```

Reviewer 2: If it's challenge-response, do we really need to be so prescriptive about all these details?

```suggestion
The signed event is an ephemeral event not meant to be published or queried. It must have a tag with the challenge string as received from the relay. Example:

```json
{
  "id": "...",
  "pubkey": "...",
  "created_at": 1669695536,
  "kind": 22242,
  "tags": [
    ["challenge", "challengestringhere"]
  ],
  "content": "",
  "sig": "..."
}
```
```

> 

Reviewer 2: A proposal for what @Semisol suggested.

```suggestion

Clients can also send an empty `["AUTH"]` message. With this, relays should end any existing authentication of the websocket and must offer a new authentication challenge.

```

before requesting `kind: 4` chat messages --, or it may do right on connection start or at some other moment it deems best. The authentication
is expected to last for the duration of the WebSocket connection.

Upon receiving a message from an unauthenticated user it can't fulfill without authentication, a relay may choose to notify the client. For
that it can use a `NOTICE` or `OK` message with a standard prefix `"restricted: "` that is readable both by humans and machines, for example:

```
["NOTICE", "restricted: we can't serve DMs to unauthenticated users, does your client implement NIP-42?"]
```

or it can return an `OK` message noting the reason an event was not written using the same prefix:

```
["OK", <event-id>, false, "restricted: we do not accept events from unauthenticated users, please sign up at https://example.com/"]
```

## Signed Event Verification

To verify `AUTH` messages, relays must ensure:

  - that the `kind` is `22242`;
  - that the event `created_at` is close (e.g. within ~10 minutes) of the current time;
  - that the `"challenge"` tag matches the challenge sent before;
  - that the `"relay"` tag matches the relay URL:

>     - URL normalization techniques can be applied. For most cases just checking if the domain name is correct should be enough.

Reviewer 2: With challenge-response we could omit prescribing all details about the event but the challenge-tag. Relays must not re-use nonces and all the rest doesn't need checking.

Lets leave it to the relays to limit the event to a reasonable size but it should really only care about the challenge-tag.

> To verify `AUTH` messages, relays must ensure:
> 
>   - that the `kind` is `22242`;
>   - that the event `created_at` is close (e.g. within ~10 minutes) of the current time;
>   - that the `"challenge"` tag matches the challenge sent before;
>   - that the `"relay"` tag matches the relay URL:
>     - URL normalization techniques can be applied. For most cases just checking if the domain name is correct should be enough.

Reviewer 2: With challenge-response we could omit prescribing all details about the event but the challenge-tag. Relays must not re-use nonces and all the rest doesn't need checking.

Lets leave it to the relays to limit the event to a reasonable size but it should really only care about the challenge-tag.

```suggestion
To verify `AUTH` messages, relays must ensure that the event's `"challenge"` tag matches the challenge sent before.
```

    before clients can query for that kind.
  - A relay may limit subscriptions of any kind to paying users or users whitelisted through any other means, and require authentication.

## Definitions

This NIP defines a new message, `AUTH`, which relays can send when they support authentication and clients can send to relays when they want
to authenticate. When sent by relays, the message is of the following form:

```
["AUTH", <challenge-string>]

> ```

Author: We can add that later if necessary.

> ```

Reviewer 2: So I would implement the challenge upon a socket being established. If the client somehow lost the nonce, it could always re-connect.

The nip is vague about when the relay might challenge indeed. Also a client might want to switch authentication, so while authed as Alice it might want to switch to auth as Bob now, so both for un-authed and authed clients there is a need to request a new nonce. Semisol's empty AUTH message triggering this has my support.

> ```

Reviewer 5: I agree and it's so simple to implement we should try to have it from the start.

> ```

Author: I think it is unnecessary bloat. How does that work on the relay side? The relay is emitting events for identity A at a given websocket, then that websocket changes to identity B, now the relay has to shove that inside the code it might have for handling the subscriptions, maybe cancel a subscription that was valid?

> ```

Author: The client should just open a new websocket if it wants a new identity -- which is also an edge case I see no use for and we shouldn't be complicating this for 99.9% of the cases.

> ```

Author: > The nip is vague about when the relay might challenge indeed

This is because it is irrelevant. All clients would just set up a listener like `when("AUTH") => send("AUTH")` and be done with it. Most relays will send the `AUTH` challenge on connect, but why specify that if it changes nothing on the code side (the client has to wait anyway)?

Also, if you think the client sending an empty `AUTH` request fixes it, consider that the relay may still take 5 minutes to reply to that with a challenge -- or may never reply, so really nothing is changed.
